schism/compiler.ss

```diff
@@ -210,33 +230,47 @@
        (define (char-ci<? c1 c2)
          (< (char->integer c1) (char->integer c2)))
        (define (list->string ls)
-         (if (pair? ls)
-             ;; For now we represent strings as lists of characters. That
-             ;; means converting between the two is just a matter of
-             ;; changing the tags.
-             (%set-tag ls ,(string-tag))
-             (error 'list->string "list->string: not a pair")))
+         (list->string-inner (make-string (length ls)) ls 0))
+       (define (list->string-inner s ls i)
+         (if (null? ls)
+             s
+             (begin
+               (string-set! s i (car ls))
+               (list->string-inner s (cdr ls) (+ 1 i)))))
        (define (string->list s)
-         (if (string? s)
-             (%set-tag s ,(pair-tag))
-             ;; calling error here can lead to an infinite loop, so we
-             ;; generate an unreachable instead.
-             (%unreachable)))
+         (string->list-inner s 0))
+       (define (string->list-inner s i)
+         (if (eq? i (string-length s))
+             '()
+             (cons (string-ref s i) (string->list-inner s (+ i 1)))))
+       (define (string-ref s i)
+         ;; TODO: make sure i is in bounds
+         (%read-mem (%as-fixnum s) (* ,(word-size) (+ i 1))))
        (define (string-equal? s1 s2)
-         (list-all-eq? (string->list s1) (string->list s2)))
-       (define (hash-chars chars h)
-         (if (null? chars)
+         (or (eq? s1 s2)
+             (and (eq? (string-length s1) (string-length s2))
+                  (string-equal-loop s1 s2 0))))
+       (define (string-equal-loop s1 s2 i)
+         (or (>= i (string-length s1))
+             (and (eq? (string-ref s1 i) (string-ref s2 i))
+                  (string-equal-loop s1 s2 (+ i 1)))))
+       (define (hash-string s i h)
+         (if (eq? i (string-length s))
              h
-             (hash-chars (cdr chars) (+ (char->integer (car chars)) h))))
+             (hash-string s (+ 1 i) (+ (char->integer (string-ref s i)) h))))
        (define (make-symbol-table n)
          (if (zero? n)
              '()
              (cons '() (make-symbol-table (- n 1)))))
-       (define (%find-symbol-by-name s ls)
-         (and (pair? ls)
-              (if (string-equal? s (car ls))
-                  (%set-tag ls ,(symbol-tag))
-                  (%find-symbol-by-name s (cdr ls)))))
+       (define (%find-symbol-by-name s table)
+         (if (or (zero? table) (null? table))
```

Comment:
I think this change to %find-symbol-by-name reverts a previous patch.  Is that intended?

Reply:
Nope, I got mixed up in the rebasing.